Suggest `collect`ing into `Vec<_>`

estebank opened the pull request:
Fix #105510.

estebank committed:
Suggest `collect`ing into `Vec<_>`

fmease commented on src/test/ui/array-slice-vec/infer_array_len.stderr:
```diff
@@ -4,7 +4,7 @@ error[E0282]: type annotations needed
    |         ^^^^^^
    |
-help: consider giving this pattern a type
+help: consider giving this pattern a type, where the placeholders `_` are specified
```
It's unfortunate that the message is kind of ambiguous here in the presence of those two other `_` inside of the slice pattern. By a beginner, it could be interpreted as if `[_, _]` needs to be adjusted, too. I don't have any suggestions though on how this could potentially be improved.

estebank replied:
I noticed that as well, but it is better than providing *no* mention of "replacing _ with a type". An alternative might be to instead of using _ as a placeholder I used `Type` as a placeholder for the case where the suggested type is `ty::Infer` and mentioned that by name.

bors commented:
:umbrella: The latest upstream changes (presumably #105525) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

estebank committed:
Remove unnecessary code and account for turbofish suggestion

Remove previously existing fallback that tried to give a good turbofish
suggestion, `need_type_info` is already good enough.

Special case `::<Vec<_>` suggestion for `Iterator::collect`.

estebank committed:
Mention implementations that satisfy the trait

estebank committed:
Suggest `: Type` instead of `: _`

compiler-errors commented on src/test/ui/array-slice-vec/infer_array_len.stderr:
```diff
@@ -4,10 +4,10 @@ error[E0282]: type annotations needed
    |
-LL |     let [_, _]: _ = a.into();
-   |               +++
+LL |     let [_, _]: Type = a.into();
```
This isn't syntactical, but I think it may be more clear the user that it's their obligation to *do* something here if we print something like `: /* Type */ =`.

That's what we do for argument mismatches (see [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7db35c152de70325381878ff2b1db174)), and IMO it's a much more obvious way to point out that this is a placeholder that the users are obligated to fill in than either `: _` or `: Type`. 

I'm just worried about users going "wait, now it's telling me that `Type` isn't a valid struct or enum!" if they just try to apply the suggestion here, as they were doing when it was suggesting `: _`.

estebank committed:
review comments

compiler-errors commented:
r? @compiler-errors i'll give this another look-over after lunch

compiler-errors commented on compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs:
```diff
@@ -183,13 +183,20 @@ fn fmt_printer<'a, 'tcx>(infcx: &'a InferCtxt<'tcx>, ns: Namespace) -> FmtPrinte
 }
 
-fn ty_to_string<'tcx>(infcx: &InferCtxt<'tcx>, ty: Ty<'tcx>) -> String {
+fn ty_to_string<'tcx>(infcx: &InferCtxt<'tcx>, ty: Ty<'tcx>, def_id: Option<DefId>) -> String {
```
This new argument really deserves a better name, it's very non-obvious what it's doing here :thinking:

compiler-errors commented on src/test/ui/inference/erase-type-params-in-label.stderr:
```diff
@@ -27,10 +27,10 @@ note: required by a bound in `bar`
    |
-LL |     let bar = bar::<T, K, Z>(1, "");
-   |                  +++++++++++
+LL |     let bar: Bar<i32, &str, Z> = bar(1, "");
```
I have no idea why this improved based on what I read in the c ode, but I'm not complaining lol

compiler-errors replied:
edit: nvm, I see why now.

compiler-errors reviewed (commented):
r=me with one nit addressed or not

estebank committed:
rename argument

estebank commented:
@bors r=compiler-errors

bors commented:
:pushpin: Commit 40a62758a7db892a47cb59de85612e8a5b9ab267 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 40a62758a7db892a47cb59de85612e8a5b9ab267 -->
<!-- homu: {"type":"Approved","sha":"40a62758a7db892a47cb59de85612e8a5b9ab267","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
